internal/grpc/services/usershareprovider/usershareprovider.go

```diff
@@ -497,40 +506,181 @@ func (s *service) GetReceivedShare(ctx context.Context, req *collaboration.GetRe
+				Status: receivedShare.GetStatus(),
+			}, err
+		}
+
+		resourceStat, err := gatewayClient.Stat(ctx, &provider.StatRequest{
+			Ref: &provider.Reference{
+				ResourceId: receivedShare.GetShare().GetShare().GetResourceId(),
+			},
+		})
+		switch {
+		case err != nil:
+			fallthrough
+		case resourceStat.GetStatus().GetCode() != rpc.Code_CODE_OK:
+			return &collaboration.UpdateReceivedShareResponse{
+				Status: resourceStat.GetStatus(),
+			}, err
+		}
+
+		// handle mount point related updates
+		{
+			// find a suitable mount point
+			var requestedMountpoint string
+			switch {
+			case slices.Contains(req.GetUpdateMask().GetPaths(), _fieldMaskPathMountPoint) && req.GetShare().GetMountPoint().GetPath() != "":
+				requestedMountpoint = req.GetShare().GetMountPoint().GetPath()
+			case receivedShare.GetShare().GetMountPoint().GetPath() != "":
+				requestedMountpoint = receivedShare.GetShare().GetMountPoint().GetPath()
+			default:
+				requestedMountpoint = resourceStat.GetInfo().GetName()
+			}
+
+			// check if the requested mount point is available and if not, find a suitable one
+			availableMountpoint, err := GetAvailableMountpoint(ctx, gatewayClient,
+				resourceStat.GetInfo().GetId(),
+				requestedMountpoint,
+			)
+			if err != nil {
+				return &collaboration.UpdateReceivedShareResponse{
+					Status: status.NewInternal(ctx, err.Error()),
+				}, nil
+			}
+
+			if !slices.Contains(req.GetUpdateMask().GetPaths(), _fieldMaskPathMountPoint) {
+				req.GetUpdateMask().Paths = append(req.GetUpdateMask().GetPaths(), _fieldMaskPathMountPoint)
+			}
+
+			req.GetShare().MountPoint = &provider.Reference{
+				Path: availableMountpoint,
+			}
+		}
+	}
+
 	var uid userpb.UserId
 	_ = utils.ReadJSONFromOpaque(req.Opaque, "userid", &uid)
-	share, err := s.sm.UpdateReceivedShare(ctx, req.Share, req.UpdateMask, &uid)
+	updatedShare, err := s.sm.UpdateReceivedShare(ctx, req.Share, req.UpdateMask, &uid)
 	if err != nil {
 		return &collaboration.UpdateReceivedShareResponse{
 			Status: status.NewInternal(ctx, "error updating received share"),
 		}, nil
 	}
 
-	res := &collaboration.UpdateReceivedShareResponse{
+	return &collaboration.UpdateReceivedShareResponse{
 		Status: status.NewOK(ctx),
-		Share:  share,
+		Share:  updatedShare,
+	}, nil
+}
+
+// GetAvailableMountpoint returns a new or existing mountpoint
+func GetAvailableMountpoint(ctx context.Context, gwc gateway.GatewayAPIClient, id *provider.ResourceId, name string) (string, error) {
+	listReceivedSharesRes, err := gwc.ListReceivedShares(ctx, &collaboration.ListReceivedSharesRequest{})
+	if err != nil {
+		return "", errtypes.InternalError("grpc list received shares request failed")
 	}
-	return res, nil
+
+	if err := errtypes.NewErrtypeFromStatus(listReceivedSharesRes.GetStatus()); err != nil {
+		return "", err
+	}
+
+	// we need to sort the received shares by mount point in order to make things easier to evaluate.
+	base := filepath.Clean(name)
+	mount := base
```

Comment:
why not rename `base` to `mount` in the first place?

Reply:
good point done